File: src/collector/hacker_news.py
import asyncio
import re
from datetime import datetime, timezone

import httpx
import structlog
from bs4 import BeautifulSoup

from src.collector.base import AbstractCollector, Article

log = structlog.get_logger()

HN_TOP_STORIES = "https://hacker-news.firebaseio.com/v0/topstories.json"
HN_ITEM = "https://hacker-news.firebaseio.com/v0/item/{id}.json"

AI_KEYWORDS = re.compile(
    r"\b(AI|LLM|GPT|ML|model|neural|machine learning|deep learning|"
    r"language model|diffusion|transformer|OpenAI|Anthropic|Gemini|Claude|"
    r"AGI|inference|fine.?tun|embedding|RAG|vector)\b",
    re.IGNORECASE,
)

MIN_SCORE = 100
MAX_AGE_SECONDS = 86400  # 24시간


class HackerNewsCollector(AbstractCollector):
    def __init__(self, timeout: float = 15.0) -> None:
        self._timeout = timeout

    async def fetch(self) -> list[Article]:
        async with httpx.AsyncClient(timeout=self._timeout) as client:
            try:
                resp = await client.get(HN_TOP_STORIES)
                resp.raise_for_status()
                story_ids: list[int] = resp.json()[:200]
            except Exception:
                log.exception("hacker_news.top_stories.failed")
                return []

            tasks = [self._fetch_item(client, sid) for sid in story_ids]
            items = await asyncio.gather(*tasks, return_exceptions=True)

        articles: list[Article] = []
        now_ts = datetime.now(timezone.utc).timestamp()

        for item in items:
            if isinstance(item, Exception) or item is None:
                continue
            if not self._is_ai_article(item, now_ts):
                continue
            articles.append(self._to_article(item))

        log.info("hacker_news.fetched", count=len(articles))
        return articles

    async def _fetch_item(self, client: httpx.AsyncClient, item_id: int) -> dict | None:
        try:
            resp = await client.get(HN_ITEM.format(id=item_id))
            resp.raise_for_status()
            return resp.json()
        except Exception:
            return None

    def _is_ai_article(self, item: dict, now_ts: float) -> bool:
        if item.get("type") != "story":
            return False
        if item.get("score", 0) < MIN_SCORE:
            return False
        title = item.get("title", "")
        if not AI_KEYWORDS.search(title):
            return False
        time_posted = item.get("time", 0)
        if now_ts - time_posted > MAX_AGE_SECONDS:
            return False
        url = item.get("url", "")
        if not url:  # Ask HN 등 외부 링크 없는 글 제외
            return False
        return True

    def _to_article(self, item: dict) -> Article:
        return Article(
            title=item.get("title", ""),
            url=item.get("url", ""),
            source="Hacker News",
            published_at=datetime.fromtimestamp(item.get("time", 0), tz=timezone.utc),
            fallback_description=item.get("title", ""),
        )


async def crawl_article_content(url: str, timeout: float = 10.0) -> str | None:
    """URL에서 본문 텍스트를 크롤링한다. 실패 시 None 반환."""
    try:
        async with httpx.AsyncClient(
            timeout=timeout,
            follow_redirects=True,
            headers={"User-Agent": "Mozilla/5.0 (compatible; article-mailer/1.0)"},
        ) as client:
            resp = await client.get(url)
            resp.raise_for_status()
            soup = BeautifulSoup(resp.text, "lxml")
            # script/style 제거
            for tag in soup(["script", "style", "nav", "footer", "header"]):
                tag.decompose()
            text = soup.get_text(separator=" ", strip=True)
            return text[:8000] if text else None
    except Exception:
        log.debug("crawl.failed", url=url)
        return None

File: src/collector/hackerNews.ts
// @mozilla/readability의 타입 정의가 DOM Document를 참조하므로 파일 스코프로 추가
/// <reference lib="dom" />
import * as cheerio from "cheerio";
import { Readability } from "@mozilla/readability";
import { parseHTML } from "linkedom";
import { AbstractCollector, Article } from "./base.js";
import logger from "../logger.js";
import { isHttpUrl } from "../utils/url.js";

const AI_KEYWORDS = ["ai", "llm", "gpt", "ml", "machine learning", "neural", "deep learning", "openai", "gemini", "claude", "anthropic", "mistral"];
const HN_API = "https://hacker-news.firebaseio.com/v0";
const MIN_SCORE = 100;
const MAX_AGE_HOURS = 24;

interface HNItem {
  id: number;
  title: string;
  url?: string;
  score: number;
  time: number;
  text?: string;
}

export async function crawlArticleContent(url: string): Promise<string | undefined> {
  if (!isHttpUrl(url)) return undefined;
  try {
    const res = await fetch(url, {
      signal: AbortSignal.timeout(10_000),
      headers: { "User-Agent": "Mozilla/5.0 (compatible; article-mailer/1.0)" },
    });
    if (!res.ok) return undefined;
    const html = await res.text();

    // Readability로 본문 추출 시도
    try {
      const { document } = parseHTML(html);
      const reader = new Readability(document as unknown as Document);
      const article = reader.parse();
      if (article?.textContent && article.textContent.trim().length > 200) {
        return article.textContent.replace(/\s+/g, " ").trim().slice(0, 8000);
      }
    } catch {
      // Readability 실패 시 cheerio fallback
    }

    // cheerio fallback
    const $ = cheerio.load(html);
    $("script, style, nav, footer, header, aside, .ad, .advertisement").remove();
    const text = $("article, main, .content, .post-body, .entry-content, body")
      .first()
      .text()
      .replace(/\s+/g, " ")
      .trim()
      .slice(0, 8000);
    return text.length > 200 ? text : undefined;
  } catch {
    return undefined;
  }
}

export class HackerNewsCollector extends AbstractCollector {
  async fetch(): Promise<Article[]> {
    const log = logger.child({ collector: "hackernews" });

    const topIds: number[] = await fetch(`${HN_API}/topstories.json`)
      .then((r) => r.json())
      .then((ids) => (ids as number[]).slice(0, 200));

    const cutoff = Date.now() / 1000 - MAX_AGE_HOURS * 3600;

    const items = await Promise.allSettled(
      topIds.map((id) =>
        fetch(`${HN_API}/item/${id}.json`).then((r) => r.json() as Promise<HNItem>)
      )
    );

    const articles: Article[] = [];
    for (const result of items) {
      if (result.status !== "fulfilled") continue;
      const item = result.value;
      if (!item.url) continue;
      if (item.score < MIN_SCORE) continue;
      if (item.time < cutoff) continue;

      const titleLower = item.title.toLowerCase();
      const isAiRelated = AI_KEYWORDS.some((kw) => titleLower.includes(kw));
      if (!isAiRelated) continue;

      articles.push({
        title: item.title,
        url: item.url,
        source: "Hacker News",
        publishedAt: new Date(item.time * 1000),
        fallbackDescription: item.text
          ? cheerio.load(item.text).text().slice(0, 300)
          : `(HN score: ${item.score}) ${item.title}`,
        score: item.score,
      });
    }

    log.info({ event: "hackernews.fetch", count: articles.length });
    return articles;
  }
}
